azdev/operations/code_gen.py

```diff
@@ -300,3 +302,61 @@ def _create_package(prefix, repo_path, is_ext, name='test', display_name=None, d
         result = pip_cmd('install -e {}'.format(new_package_path), "Installing `{}{}`...".format(prefix, name))
         if result.error:
             raise result.error  # pylint: disable=raising-bad-type
+
+
+def _get_swagger_readme_file_path(ext_name, swagger_repo):
+    from urllib import request, error
+
+    swagger_readme_file_path = None
+    if swagger_repo == const.GITHUB_SWAGGER_REPO_URL or \
+            (swagger_repo.startswith('https://') and swagger_repo.endswith(const.SWAGGER_REPO_NAME)):
+        swagger_readme_file_path = '{}/blob/master/specification/{}/resource-manager'.format(
+            swagger_repo, ext_name)
+        # validate URL
+        try:
+            request.urlopen(swagger_readme_file_path)
+        except error.HTTPError as ex:
+            raise CLIError('HTTPError: {}\nThe URL {} does not exist.'.format(ex.code, swagger_readme_file_path))
+    else:
+        swagger_readme_file_path = os.path.join(swagger_repo, 'specification', ext_name, 'resource-manager')
+        if not os.path.isdir(swagger_readme_file_path):
+            raise CLIError("The path {} does not exist.".format(swagger_readme_file_path))
+    return swagger_readme_file_path
+
+
+def _generate_extension(ext_name, repo_path, swagger_readme_file_path, use):
+    heading('Start generating extension {}.'.format(ext_name))
+    # check if npm is installed
+    try:
+        subprocess.run('npm --version', shell=True, check=True, stdout=subprocess.DEVNULL)
+    except subprocess.CalledProcessError as ex:
+        raise CLIError('{}\nPlease install npm.'.format(ex))
+    # check if autorest is installed
+    try:
+        subprocess.run(
+            'npm list -g autorest', shell=True, check=True, stdout=subprocess.DEVNULL, stderr=subprocess.STDOUT)
+    except subprocess.CalledProcessError:
+        display('Installing autorest.\n')
+        try:
+            if const.IS_WINDOWS:
+                subprocess.run('npm install -g autorest', shell=True, check=True)
+            else:
+                display('May need to provide your password.\n')
+                subprocess.run('sudo npm install -g autorest', shell=True, check=True)
```

Comment:
why need `sudo` permission?

Reply:
It needs admin permission when installing autorest in Linux.

Reply:
any option to install without permission?

Reply:
Requireing user to install via sudo would be too much.

Reply:
There has a workaround to do global npm install without sudo, but will make nvm lost node version control.

Reply:
could you let user run `sudo azdev extension add` instead of this way?